llap-tez/src/java/org/apache/hadoop/hive/llap/tezplugins/LlapTaskSchedulerService.java

```diff
@@ -3154,4 +3165,28 @@ public void taskInfoUpdated(TezTaskAttemptID attemptId, boolean isGuaranteed) {
         + attemptId + ", " + newState);
     sendUpdateMessageAsync(ti, newState);
   }
+
+  private void updateMetrics(TaskAttemptImpl taskAttempt) {
+    // Only do it for map tasks
+    if (!isMapTask(taskAttempt)) {
+      return;
+    }
+    // Check if this task was already assigned to a node
+    NodeInfo nodeInfo = knownTasks.get(taskAttempt).assignedNode;
+    if (nodeInfo == null) {
+      return;
+    }
+
+    metrics.addTaskLatency(nodeInfo.shortStringBase, taskAttempt.getFinishTime() - taskAttempt.getLaunchTime());
+  }
+
+  private boolean isMapTask(TaskAttemptImpl taskAttempt) {
+    boolean isMapTask = false;
+    for(TezCounter counter : taskAttempt.getCounters().getGroup("HIVE")) {
+      if(counter.getName().startsWith("RECORDS_IN_Map")) {
+        isMapTask = true;
```

Comment:
....and if you prefer to iterate the group, you should break here instead of comparing all counters to the end.

Reply:
Optimized the counter based solution as suggested.